src/vm/comutilnative.cpp

```diff
@@ -1416,20 +1417,45 @@ FCIMPL5(VOID, Buffer::BlockCopy, ArrayBase *src, int srcOffset, ArrayBase *dst,
     if (src==NULL || dst==NULL)
         FCThrowArgumentNullVoid((src==NULL) ? W("src") : W("dst"));
 
-    // Size of the Arrays in bytes
-    SIZE_T srcLen = src->GetNumComponents() * src->GetComponentSize();
-    SIZE_T dstLen = srcLen;
+    SIZE_T srcLen, dstLen;
 
-    // We only want to allow arrays of primitives, no Objects.
-    const CorElementType srcET = src->GetArrayElementType();
-    if (!CorTypeInfo::IsPrimitiveType_NoThrow(srcET))
-        FCThrowArgumentVoid(W("src"), W("Arg_MustBePrimArray"));
+    //
+    // Use specialized fast path for byte arrays because of it is what Buffer::BlockCopy is 
+    // typically used for.
+    //
 
-    if (src != dst) {
-        const CorElementType dstET = dst->GetArrayElementType();
-        if (!CorTypeInfo::IsPrimitiveType_NoThrow(dstET))
-            FCThrowArgumentVoid(W("dest"), W("Arg_MustBePrimArray"));
-        dstLen = dst->GetNumComponents() * dst->GetComponentSize();
+    MethodTable * pByteArrayMT = g_pByteArrayMT;
+    if (pByteArrayMT == NULL)
+    {
+        if (g_pPredefinedArrayTypes[ELEMENT_TYPE_U1] != NULL)
```

Comment:
In what situations will this be null? Just curious, as until it's not null it looks like we won't take the fast path.